rosbag2_transport/src/rosbag2_transport/topic_filter.cpp

```diff
@@ -129,36 +149,91 @@ bool TopicFilter::take_topic(
+    {
+      return false;
+    }
 
-  std::regex exclude_regex(record_options_.exclude);
-  if (!record_options_.exclude.empty() && std::regex_search(topic_name, exclude_regex)) {
-    return false;
-  }
+    if (!record_options_.all_topics) {
+      // Not in include topic list
+      if (record_options_.topics.empty() || !topic_in_list(topic_name, record_options_.topics)) {
+        // Not match include regex
+        if (!record_options_.regex.empty()) {
+          std::regex include_regex(record_options_.regex);
+          if (!std::regex_search(topic_name, include_regex)) {
+            return false;
+          }
+        } else {
+          return false;
+        }
+      }
+    }
 
-  std::regex include_regex(record_options_.regex);
-  if (
-    !record_options_.all &&  // All takes precedence over regex
-    !record_options_.regex.empty() &&  // empty regex matches nothing, but should be ignored
-    !std::regex_search(topic_name, include_regex))
-  {
-    return false;
-  }
+    if (!record_options_.exclude_topics.empty() &&
+      topic_in_list(topic_name, record_options_.exclude_topics))
+    {
+      return false;
+    }
 
-  if (!has_single_type(topic_name, topic_types)) {
-    return false;
-  }
+    if (!record_options_.exclude_regex.empty()) {
+      std::regex exclude_regex(record_options_.exclude_regex);
+      if (std::regex_search(topic_name, exclude_regex)) {
+        return false;
+      }
+    }
 
-  if (!record_options_.include_hidden_topics && topic_is_hidden(topic_name)) {
-    RCUTILS_LOG_WARN_ONCE_NAMED(
-      ROSBAG2_TRANSPORT_PACKAGE_NAME,
-      "Hidden topics are not recorded. Enable them with --include-hidden-topics");
-    return false;
+    if (!record_options_.include_hidden_topics && topic_is_hidden(topic_name)) {
+      RCUTILS_LOG_WARN_ONCE_NAMED(
+        ROSBAG2_TRANSPORT_PACKAGE_NAME,
+        "Hidden topics are not recorded. Enable them with --include-hidden-topics");
+      return false;
+    }
+  } else {
+    if (!record_options_.all_services &&
+      record_options_.services.empty() &&
+      record_options_.regex.empty())
+    {
+      return false;
+    }
+
+    // Convert service event topic name to service name
+    auto service_name = rosbag2_cpp::service_event_topic_name_to_service_name(topic_name);
+
+    if (!record_options_.all_services) {
+      // Not in include service list
+      if (record_options_.services.empty() ||
+        !service_in_list(topic_name, record_options_.services))
+      {
+        // Not match include regex
+        if (!record_options_.regex.empty()) {
+          std::regex include_regex(record_options_.regex);
+          if (!std::regex_search(service_name, include_regex)) {
+            return false;
+          }
+        } else {
+          return false;
+        }
+      }
+    }
+
+    if (!record_options_.exclude_service_events.empty() &&
+      service_in_list(topic_name, record_options_.exclude_service_events))
```

Comment:
```suggestion
      service_in_list(service_name, record_options_.exclude_service_events))
```

Reply:
record_options_.exclude_service_events saved service event topic list.   
So topic_name is used.   
The name of this function is not well-chosen.